File: core-android/src/main/java/com/lookaround/core/android/map/clustering/ClusterRenderer.java
package com.lookaround.core.android.map.clustering;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.mapzen.tangram.LngLat;
import com.mapzen.tangram.MapController;
import com.mapzen.tangram.Marker;
import com.mapzen.tangram.MarkerPickListener;
import com.mapzen.tangram.MarkerPickResult;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class ClusterRenderer<T extends ClusterItem> implements MarkerPickListener {
    private final MapController mapController;

    private final List<Cluster<T>> mClusters = new ArrayList<>();

    private final Map<Cluster<T>, Marker> mMarkers = new HashMap<>();

    private IconGenerator<T> mIconGenerator;

    private ClusterManager.Callbacks<T> mCallbacks;

    private final Context context;

    ClusterRenderer(@NonNull Context context, @NonNull MapController mapController) {
        this.mapController = mapController;
        this.mapController.setMarkerPickListener(this);
        this.context = context;
        mIconGenerator = new DefaultIconGenerator<>(context);
    }

    void setCallbacks(@Nullable ClusterManager.Callbacks<T> listener) {
        mCallbacks = listener;
    }

    void setIconGenerator(@NonNull IconGenerator<T> iconGenerator) {
        mIconGenerator = iconGenerator;
    }

    void render(@NonNull List<Cluster<T>> clusters) {
        List<Cluster<T>> clustersToAdd = new ArrayList<>();
        List<Cluster<T>> clustersToRemove = new ArrayList<>();

        for (Cluster<T> cluster : clusters) {
            if (!mMarkers.containsKey(cluster)) {
                clustersToAdd.add(cluster);
            }
        }

        for (Cluster<T> cluster : mMarkers.keySet()) {
            if (!clusters.contains(cluster)) {
                clustersToRemove.add(cluster);
            }
        }

        mClusters.addAll(clustersToAdd);
        mClusters.removeAll(clustersToRemove);

        for (Cluster<T> clusterToRemove : clustersToRemove) {
            Marker markerToRemove = mMarkers.get(clusterToRemove);
            if (markerToRemove == null) continue;
            mapController.removeMarker(markerToRemove);

            mMarkers.remove(clusterToRemove);
        }

        for (Cluster<T> clusterToAdd : clustersToAdd) {
            Marker markerToAdd = mapController.addMarker();
            markerToAdd.setPoint(new LngLat(clusterToAdd.getLongitude(), clusterToAdd.getLatitude()));
            markerToAdd.setStylingFromString(
                    "{ style: 'points', size: [27px, 27px], order: 2000, collide: false, color: blue}"
            );
            Bitmap markerIcon = getMarkerIcon(clusterToAdd);
            markerToAdd.setDrawable(new BitmapDrawable(context.getResources(), markerIcon));
            markerToAdd.setVisible(true);
            markerToAdd.setUserData(clusterToAdd);

            mMarkers.put(clusterToAdd, markerToAdd);
        }
    }

    @NonNull
    private Bitmap getMarkerIcon(@NonNull Cluster<T> cluster) {
        Bitmap clusterIcon;

        List<T> clusterItems = cluster.getItems();
        if (clusterItems.size() > 1) {
            clusterIcon = mIconGenerator.getClusterIcon(cluster);
        } else {
            clusterIcon = mIconGenerator.getClusterItemIcon(clusterItems.get(0));
        }

        return Preconditions.checkNotNull(clusterIcon);
    }

    @Override
    public void onMarkerPickComplete(@Nullable MarkerPickResult markerPickResult) {
        if (markerPickResult == null || markerPickResult.getMarker() == null) return;

        Object markerTag = markerPickResult.getMarker().getUserData();
        if (markerTag instanceof Cluster) {
            //noinspection unchecked
            Cluster<T> cluster = (Cluster<T>) markerPickResult.getMarker().getUserData();
            List<T> clusterItems = cluster.getItems();
            if (mCallbacks != null) {
                if (clusterItems.size() > 1) {
                    mCallbacks.onClusterClick(cluster);
                } else {
                    mCallbacks.onClusterItemClick(clusterItems.get(0));
                }
            }
        }
    }
}

File: core-android/src/main/java/com/lookaround/core/android/map/clustering/ClusterRenderer.kt
package com.lookaround.core.android.map.clustering

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import com.lookaround.core.android.ext.MarkerPickResult
import com.lookaround.core.android.ext.TangramMarkerPickResult
import com.lookaround.core.android.ext.latLon
import com.lookaround.core.android.map.model.LatLon
import com.mapzen.tangram.LngLat
import com.mapzen.tangram.MapController
import com.mapzen.tangram.Marker
import kotlinx.coroutines.ExperimentalCoroutinesApi
import timber.log.Timber

@ExperimentalCoroutinesApi
internal class ClusterRenderer<T : ClusterItem>(
    private val context: Context,
    private val mapController: MapController,
    private val iconGenerator: IconGenerator<T> = DefaultIconGenerator(context)
) {
    private val clusters: MutableList<Cluster<T>> = ArrayList()
    private val markers: MutableMap<Cluster<T>, Marker> = HashMap()

    fun onMarkerPickComplete(markerPickResult: TangramMarkerPickResult?): MarkerPickResult? {
        val marker = markerPickResult?.marker ?: return null
        @Suppress("UNCHECKED_CAST") val cluster = marker.userData as? Cluster<T> ?: return null
        val clusterItems = cluster.items
        return if (clusterItems.size > 1) {
            MarkerPickResult(markerPickResult.coordinates.latLon)
        } else {
            val item = clusterItems[0]
            MarkerPickResult(LatLon(item.latitude, item.longitude), item.uuid)
        }
    }

    fun render(clusters: List<Cluster<T>>) {
        val clustersToAdd = mutableListOf<Cluster<T>>()
        val clustersToRemove = mutableListOf<Cluster<T>>()

        for (cluster in clusters) {
            if (!markers.containsKey(cluster)) clustersToAdd.add(cluster)
        }

        for (cluster in markers.keys) {
            if (!clusters.contains(cluster)) clustersToRemove.add(cluster)
        }

        this.clusters.addAll(clustersToAdd)
        this.clusters.removeAll(clustersToRemove)

        for (clusterToRemove in clustersToRemove) {
            val markerToRemove = markers[clusterToRemove] ?: continue
            mapController.removeMarker(markerToRemove)
            try {
                markers.remove(clusterToRemove)
            } catch (throwable: Throwable) {
                Timber.e(throwable)
            }
        }

        for (clusterToAdd in clustersToAdd) {
            markers[clusterToAdd] =
                mapController.addMarker().apply {
                    setPoint(LngLat(clusterToAdd.longitude, clusterToAdd.latitude))
                    setStylingFromString(
                        "{ style: 'points', size: [27px, 27px], order: 2000, flat: true, collide: false, color: 'white', interactive: true}"
                    )
                    val markerIcon = getMarkerIcon(clusterToAdd)
                    setDrawable(BitmapDrawable(context.resources, markerIcon))
                    setDrawOrder(10)
                    isVisible = true
                    userData = clusterToAdd
                }
        }
    }

    private fun getMarkerIcon(cluster: Cluster<T>): Bitmap {
        val clusterItems = cluster.items
        return if (clusterItems.size > 1) iconGenerator.getClusterIcon(cluster)
        else iconGenerator.getClusterItemIcon(clusterItems[0])
    }
}
